### DISCUSSION TITLE
Support DSPy

### DISCUSSION THREAD

**Discussion by sandangel** [langfuse/langfuse#1295] [Ideas]
Hi, it would be great if we can add support for DSPy, a much better approach than langchain prompt template.

**Comment by marcklingen** [comment]
Tinkered a bit with DSPy and it seems promising. Do you use it for a production app right now?

Let's see how much interest there is for this on this idea post.

**Reply by BlueKiji77** [reply]
Is there a straightforwardway to use langfuse with dspy like with langchain and llamaindex?

**Reply by marcklingen** [reply]
Did not get to trying to build an integration or cookbook yet. I'd really appreciate any contribution on this if you want to have a deeper look and come up with something. Happy to support if you have any questions related to the low-level python sdk

**Comment by marcklingen** [comment]
+1 by @sutyum in #1300

**Comment by sandangel** [comment]
Please take a look at how we could do function call with DSPy: https://gist.github.com/seanchatmangpt/e6fae3343f32c50f5a07e2ffe2e8b0f3.

Exceptionally well-organized. With langfuse integration and tracing ability, it would be perfect.

**Comment by GMartin-dev** [comment]
I also agree that most projects will use DSPy sooner or later, at least for some prompts.
There is also mix support of DSPy for langchain runnables, so they are not mutually exclusive.
https://python.langchain.com/docs/integrations/providers/dspy
You can start with a handmade prompt to get the app working quickly then you start optimizing with DSPy.
Even if you start with DSPy sooner or later you need to manually tweaks prompts later, perhaps that's not the case for simpler prompts but when you have a complex orchestration with a lot of moving parts you will end having a mix.
DSPy is casually promoting arize-phoenix for observability but it seems like a tool for investigation not for production apps.
So, in my humble view, if langfuse can support, for example, langchain runnables and DSPy concurrently  it could become a default solution for complex projects.

**Comment by wdhitchc** [comment]
Adding my +1. DSPy is amazing, I use it in multiple production apps. Langfuse + DSPY could play so well together.

**Comment by SinghCoder** [comment]
Adding my +1 also.

**Comment by sfarthin** [comment]
I also think DSPy can be incredibly useful, but I think LangFuse needs a better higher-level abstraction so we can plug in a number of different implementations and approaches. See https://github.com/orgs/langfuse/discussions/1664

**Comment by paolotamag** [comment]
any news on the DSPy integration? or is the decorator the only way to send traces to langfuse at this time? 
it seems DSPy is creating the prompts for me, so I would need to extrapolate those first, and then give those to langfuse?
seems like extra work.. maybe it's better if I wait for an integration?

**Reply by paolotamag** [reply]
like how the hack i can get the full trace generate by this thing
https://dspy-docs.vercel.app/api/modules/ChainOfThought

or is that cut-off "rationale" all the system prompt the LLM gets?

**Reply by paolotamag** [reply]
you can find the generated prompts with 'lm.history[-1]['prompt']' where 'lm = dspy.OpenAI(model=self.model, max_tokens=self.max_tokens, temperature=0)'. So my understanding is that first you fire off a signature then use that to retrieve what to send to langfuse...

**Comment by marcklingen** [comment]
I’ve just added a notebook featuring the Langfuse integration that was added to dspy: https://langfuse.com/docs/integrations/dspy
